Add EIP-4824: Common Interfaces for DAOs (#4824)
ethereum/EIPs: EIPS/eip-4824.md

---
eip: 4824

> title: DAO API

Reviewer: I'm still not sold on this title... I know it's been bikeshedded a lot, but:

 - Proposal, Membership, Activity, and Governance Interfaces for DAOs
 - DAO Metadata Interface

Author: It's been a long, long conversation 😅  the other co-authors and I went through a bunch of alternatives, had our own disagreements about we felt made the most sense, and we came out with DAO API.

Re: "DAO Metadata Interface", the point is that not all the data is "metadata". Would you call proposals data being updated and consumed on a daily basis the "metadata" of the DAO?

Thanks again for all your comments!

Reviewer: While I don't have a clearly better suggestion, `DAO API` isn't descriptive enough for an EIP.

Here's a couple more random ones (none of them great):

 * Proposal, Membership, Activity, and Governance Interfaces for DAOs
 * Operational Interfaces for DAOs
 * Common Interfaces for DAOs
 * Universal Interfaces for DAOs

The first one is the most descriptive, but it's quite long.

Author: Yeah, I'm not sure how to respond. If it is the name that is holding up publishing of this EIP, at this point we are happy to change it to whatever the EIP editors want to call it.

Reviewer 2: The name is definitely holding off the publishing of the EIP. I would also prefer the first name:
```suggestion
title: Proposal, Membership, Activity, and Governance Interfaces for DAOs
```

Author: Took a vote of the co-authors, and of these options, we preferred "Common Interfaces for DAOs". I've implemented that change.

We are wary of the longer name because it takes the form of a list, and there's a good chance that the elements of the list may need to change or be extended in the future, which daoURI is intended to support.

> title: Decentralized Autonomous Organizations

Reviewer 3: Suggest updating the name to `Metadata Standard of Decentralized Autonomous Organizations` to reflect the general scope of this proposal

Author: See response below on your other name change suggestion!

Reviewer 4: @thelastjosh this title will likely cause confusion among many people. I would like to request that you and your coauthors come up with a more specific title regarding the proposal.

Author: @lightclient Changed.

description: A standard URI and JSON schema for decentralized autonomous organizations (DAOs).
author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)
discussions-to: https://ethereum-magicians.org/t/eip-4824-decentralized-autonomous-organizations/8362
status: Draft
type: Standards Track
category: ERC
created: 2022-02-17
---

## Abstract

> A standard URI and JSON schema for decentralized autonomous organizations (DAOs), focusing on relating on-chain and off-chain representations of membership and proposals.

Reviewer 3: I think this is one of the first proposals using JSON schema as the main format. 

Consider using Solidity or [ABI](https://docs.soliditylang.org/en/develop/abi-spec.html) or provide your rationale to justify starting a new precedent?

Author: It doesn't make sense to use Solidity to store or define off-chain data which we would want to convert to JSON or some similar format in any case. As for why we decided not to standardize more on-chain data, we state in the rationale section:

> DAOs themselves have a number of existing and emerging use-cases. But almost all DAOs need to publish data off-chain for a number of reasons: communicating to and recruiting members, coordinating activities, powering user interfaces and governance applications such as Snapshot or Tally, or enabling search and discovery via platforms like DeepDAO or Messari. Having a standardized schema for this data, akin to an API specification, would strengthen existing use-cases for DAOs, help scale tooling and frameworks across the ecosystem, and build support for additional forms of interoperability.
> 
> While we considered standardizing on-chain aspects of DAOs in this standard, particularly on-chain proposal objects and proposal IDs, we felt that this level of standardization was premature given (1) the relative immaturity of use-cases, such as multi-DAO proposals or master-minion contracts, that would benefit from such standardization, (2) the close linkage between proposal systems and governance, which we did not want to standardize (see “governanceURI”, below), and (3) the prevalence of off-chain and L2 voting and proposal systems in DAOs (see “proposalsURI”, below). Further, a standard URI interface is relatively easy to adopt and has been actively demanded by frameworks (see “Community Consensus”, below).

## Motivation

> DAOs, since being invoked in the Ethereum whitepaper, have been vaguely defined. This has led to a wide range of patterns but little standardization or interoperability between the frameworks and tools that have emerged. Standardization and interoperability are necessary to support a variety of use-cases. In particular, a standard daoURI, similar to tokenURI in [ERC-721](/EIPS/eip-721), will enhance DAO search, discoverability, legibility, and proposal simulation. More consistent data across the ecosystem is also a prerequisite for future DAO standards.

Reviewer 3: For discoverability, consider incorporate [EIP-165](https://eips.ethereum.org/EIPS/eip-165)?

Author: Hi @xinbenlv , thanks for the comments! Sorry for the delay in replying; I wanted to have our next working group meeting so I could give others a chance to review some of your suggestions. 

Re: using EIP-165: yes! Thanks for catching this. I believe we had even discussed EIP-165 early on in the meetings but forgot to come back to it before the draft was finalized.

Author: BTW, I'll just add we're trying to encourage discussions at https://ethereum-magicians.org/t/erc-4824-decentralized-autonomous-organizations/8362 in case you have additional feedback!

> DAOs, since being invoked in the Ethereum whitepaper, have been vaguely defined. This has led to a wide range of patterns but little standardization or interoperability between the frameworks and tools that have emerged. Standardization and interoperability are necessary to support a variety of use-cases. In particular, a standard daoURI, similar to tokenURI in [ERC-721](/EIPS/eip-721), will enhance DAO search, discoverability, legibility, and proposal simulation. More consistent data across the ecosystem is also a prerequisite for future DAO standards.

Reviewer 4: ```suggestion
DAOs, since being invoked in the Ethereum whitepaper, have been vaguely defined. This has led to a wide range of patterns but little standardization or interoperability between the frameworks and tools that have emerged. Standardization and interoperability are necessary to support a variety of use-cases. In particular, a standard daoURI, similar to tokenURI in [ERC-721](./eip-721), will enhance DAO search, discoverability, legibility, and proposal simulation. More consistent data across the ecosystem is also a prerequisite for future DAO standards.
```

Reviewer 4: Please use relative paths to all EIPs.

## Specification

> The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

Reviewer 3: This can be left out because they are already defined in EIP-1

Author: I believe EIP-1 specifically encourages us to include this section in every EIP.

> EIPs are encouraged to follow [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)[](https://eips.ethereum.org/EIPS/eip-1#history) for terminology and to insert the following at the beginning of the Specification section:
> 
> The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> Every EIP-4824 compliant contract MUST implement the EIP4824 interface below:

Reviewer: ```suggestion
Every EIP-4824 compliant contract MUST implement the `EIP4824` interface below:
```

```solidity
pragma solidity ^0.4.20;

/// @title EIP-4824 DAOs
/// @dev See https://eips.ethereum.org/EIPS/eip-4824

interface EIP4824 {
    /// @notice A distinct Uniform Resource Identifier (URI) pointing to a JSON object following the "EIP-4824 DAO JSON-LD Schema". This JSON file splits into four URIs: membersURI, proposalsURI, activityLogURI, and governanceURI. The membersURI should point to a JSON file that conforms to the "EIP-4824 Members JSON-LD Schema". The proposalsURI should point to a JSON file that conforms to the "EIP-4824 Proposals JSON-LD Schema". The activityLogURI should point to a JSON file that conforms to the "EIP-4824 Activity Log JSON-LD Schema". The governanceURI should point to a flatfile, normatively a .md file. Each of the JSON files named above can be statically-hosted or dynamically-generated.

>     function daoURI() external view returns (string _daoURI);

Reviewer 3: Does it mean the caller of contract will just call `EIP4824 .daoURI` and then get a JSON schema that has other? 

Can you elaborate if considered making all the metadata available in the form of a Solidity `interface`? eg ERC-721

Reviewer 3: Or if this is intentional that daoURI will return offchain information in JSON-schema, maybe update the name of this proposal to `Off-Chain metadata standard for DAOs`

Author: > Can you elaborate if considered making all the metadata available in the form of a Solidity interface? eg ERC-721
Interesting thought. To be clear, you're suggesting we make the `name` and `description` fields (but probably not `type`) available directly in the Solidity interface. We talked a bit about standardizing certain data on-chain, mainly around proposals, but never discussed standardizing these metadata fields directly on-chain.

My initial thought here is that having this basic "storefront" data in the JSON object covers the legibility, discoverability, and reporting use-cases that we want to support, and that it is marginally more elegant and easier for the DAO end-user just to have to set daoURI in the contract interface itself. I can kind of imagine use-cases where a contract just wants the name and description without having to call daoURI / interact with the outside world, but generally speaking the off-chain data will be fresher. Side note: ERC-721 puts just `name` and `symbol` in the Solidity interface while putting `name` and `description` in the off-chain JSON schema, so it's also not totally consistent or clear what fields should end up on-chain vs off-chain. Do you have any idea why they made those decisions?

Author: Re: proposed name change. We'd prefer to keep the current title.

As the EIP states, DAOs have been vaguely defined. 10 experts walk into a room with 10 definitions of what a DAO is. One of the things that came out of the cross-industry conversations (see the 'Community Consensus' section) that generated the standard was the urgent need for something definitive for the ecosystem. So part of the challenge was coming up with a definitive standard for DAOs (and not just off-chain metadata standards for DAOs) without over-specifying them.

}
```

The EIP-4824 DAO JSON-LD Schema mentioned above.

```json
{

> 	"@context": "http://www.daostar.org/schemas",

Reviewer: I'm not sure what this is (and the link 404s), but would it make sense to include whatever is supposed to be there in the `assets` directory of this EIP instead?

Author: Looks like the redirect hasn't been set up; that should get fixed this week! 

I think keeping a snapshot of the context in the assets folder is a good idea. That said, the context files / the ontology will need to grow & change with additional API extensions built on top of this standard.

Reviewer 2: > I think keeping a snapshot of the context in the assets folder is a good idea. That said, the context files / the ontology will need to grow & change with additional API extensions built on top of this standard.

Those extensions can provide their own schema in their own `assets` directory.

Author: Added the current schema to the assets directory.

Reviewer 3: Can we remove`http://www.daostar.org/` per https://github.com/ethereum/EIPs/blob/master/EIPS/eip-1.md#linking-to-external-resources?

	"type": "DAO",
	"name": "<name of the DAO>",
	"description": "<description>",
	"membersURI": "<URI>",
	"proposalsURI": "<URI>",
	"activityLogURI": "<URI>",
	"governanceURI": "<URI>"
}

> ```

Reviewer: ERC-721 uses [JSON Schema](https://json-schema.org/) to define its metadata structure. Would you be opposed to using that here?

Reviewer: I wouldn't mind either JSON schema or JSON-LD. I'm mostly just concerned with consistent and usable standards. Either way, the validation rules probably should be in this EIP, so that an implementer doesn't need any external resources.

### **Community Consensus**

The initial draft standard was developed as part of the DAOstar One roundtable series, which included representatives from all major EVM-based DAO frameworks (Aragon, Compound, DAOstack, Gnosis, Moloch, OpenZeppelin, and Tribute), a wide selection of DAO tooling developers, as well as several major DAOs. Thank you to all the participants of the roundtable, the full list of which can be found [here](https://daostar.one/). We would especially like to thank Auryn Macmillan, Fabien of Snapshot, Selim Imoberdorf, Lucia Korpas, and Mehdi Salehi for their contributions.

In-person events will be held at Schelling Point 2022 and at ETHDenver 2022, where we hope to receive more comments from the community. We also plan to schedule a series of community calls through early 2022.

> 

Reviewer 4: ```suggestion

## Backwards Compatibility
TBD

```

Reviewer 4: required section

## Security Considerations

This standard defines the interfaces for the DAO URIs but does not specify the rules under which the URIs are set, or how the data is prepared. Developers implementing this standard should consider how to update this data in a way aligned with the DAO’s governance model, and keep the data fresh in a way that minimizes reliance on centralized service providers.

Indexers that rely on the data returned by the URI should take caution if DAOs return executable code from the URIs. This executable code might be intended to get the freshest information on membership, proposals, and activity log, but it could also be used to run unrelated tasks.

## Copyright

> Copyright and related rights waived via [CC0](https://creativecommons.org/publicdomain/zero/1.0/).

Reviewer 2: ```suggestion
Copyright and related rights waived via [CC0](../LICENSE.md).
```

---
eip: 4824
title: Common Interfaces for DAOs

> description: An API standard for decentralized autonomous organizations (DAOs).

Reviewer 2: ```suggestion
description: An API for decentralized autonomous organizations (DAOs).
```
Shouldn't contain the word `standard`

> author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)

Author: Weird :D done!

> author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)

Reviewer 5: Hmm, Code Spell shouldn't be failing on people's names.  If it is we should fix that, as it isn't reasonable to add everyone's name to the whitelist I don't think.  Also, I have never seen it fail on a name before.

> author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)

Reviewer 5: Just dug into this a bit.  I guess the way codespell works is that it looks for words that are "likely misspelled" rather than checking a dictionary.  Normally, names are sufficiently unrelated to words that codespell doesn't detect them as misspellings of other words, but in this case the name is too close to an English word.

I wonder if there is some way we can add a temporary pass without having to permanently add the name to the dictionary?

> author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)

Reviewer 2: > I wonder if there is some way we can add a temporary pass without having to permanently add the name to the dictionary?

... remove codespell.

> author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)

Reviewer 5: It is quite useful, it catches a *lot* of typos.

> author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)

Author: I can also ask Sam @nivida to change his name...

> author: Joshua Tan (@thelastjosh), Isaac Patka (@ipatka), Ido Gershtein <ido@daostack.io>, Eyal Eithcowich <eyal@deepdao.io>, Michael Zargham (@mzargham), Sam Furter (@nivida)

Author: @MicahZoltu @Pandapip1 Hi folks, I think this is the only thing holding back publishing; we're already starting to roll out additional support + integrating with Moloch v2 & v3. Would love your help getting this through 🙏

discussions-to: https://ethereum-magicians.org/t/eip-4824-decentralized-autonomous-organizations/8362
status: Draft
type: Standards Track
category: ERC
created: 2022-02-17
---

## Abstract

An API standard for decentralized autonomous organizations (DAOs), focused on relating on-chain and off-chain representations of membership and proposals.

## Motivation

> DAOs, since being invoked in the Ethereum whitepaper, have been vaguely defined. This has led to a wide range of patterns but little standardization or interoperability between the frameworks and tools that have emerged. Standardization and interoperability are necessary to support a variety of use-cases. In particular, a standard daoURI, similar to tokenURI in [ERC-721](./eip-721), will enhance DAO discoverability, legibility, proposal simulation, and interoperability between tools. More consistent data across the ecosystem is also a prerequisite for future DAO standards.

Reviewer 2: ```suggestion
DAOs, since being invoked in the Ethereum whitepaper, have been vaguely defined. This has led to a wide range of patterns but little standardization or interoperability between the frameworks and tools that have emerged. Standardization and interoperability are necessary to support a variety of use-cases. In particular, a standard daoURI, similar to tokenURI in [EIP-721](./eip-721), will enhance DAO discoverability, legibility, proposal simulation, and interoperability between tools. More consistent data across the ecosystem is also a prerequisite for future DAO standards.
```

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

> Every EIP-4824 compliant contract MUST implement the `EIP4824` interface below:

Reviewer 2: ```suggestion
Every contract implementing this EIP MUST implement the `EIP4824` interface below:
```

```solidity
pragma solidity ^0.4.20;

/// @title EIP-4824 DAOs
/// @dev See https://eips.ethereum.org/EIPS/eip-4824

interface EIP4824 {
    /// @notice A distinct Uniform Resource Identifier (URI) pointing to a JSON object following the "EIP-4824 DAO JSON-LD Schema". This JSON file splits into four URIs: membersURI, proposalsURI, activityLogURI, and governanceURI. The membersURI should point to a JSON file that conforms to the "EIP-4824 Members JSON-LD Schema". The proposalsURI should point to a JSON file that conforms to the "EIP-4824 Proposals JSON-LD Schema". The activityLogURI should point to a JSON file that conforms to the "EIP-4824 Activity Log JSON-LD Schema". The governanceURI should point to a flatfile, normatively a .md file. Each of the JSON files named above can be statically-hosted or dynamically-generated.
    function daoURI() external view returns (string _daoURI);
}
```

> The EIP-4824 DAO JSON-LD Schema mentioned above.

Reviewer 2: ```suggestion
The DAO JSON-LD Schema mentioned above:
```

```json
{
	"@context": "http://www.daostar.org/schemas",
	"type": "DAO",
	"name": "<name of the DAO>",
	"description": "<description>",
	"membersURI": "<URI>",
	"proposalsURI": "<URI>",
	"activityLogURI": "<URI>",
	"governanceURI": "<URI>"
}
```

> A DAO MAY inherit the EIP4824 interface above or it MAY create an external registration contract that is compliant with EIP-4824. The external registration contract MUST store the DAO’s primary address.

Reviewer 2: ```suggestion
A DAO MAY inherit the above interface above or it MAY create an external registration contract that is compliant with this EIP. The external registration contract MUST store the DAO’s primary address.
```

		function initialize(address _daoAddress) {
			if (_daoAddress == msg.sender) {
				daoAddress = _daoAddress;
				emit TransferAccepted(address(0), _daoAddress);
			} else {
				offeredAddress = _daoAddress;
				emit TransferOffered(address(0), _daoAddress);
			}

		}

		function daoURI() external view returns (string _daoURI) {
			return daoURI;
		}
		
    function offerTransfer(address _offeredAddress) external {
			if (msg.sender != daoAddress) revert NotOwner();
			offeredAddress = _offeredAddress;

			emit TransferOffered(daoAddress, _offeredAddress);
		}

    function acceptTransfer() external {
			address _oldOwner = daoAddress;
			if (msg.sender != offeredAddress) revert NotOffered();
			daoAddress = msg.sender;
			emit TransferAccepted(_oldOwner, daoAddress);
		}
}
```

If a DAO uses an external registration contract, the DAO SHOULD use a common registration factory contract to enable efficient network indexing.

```solidity
pragma solidity ^0.8.1;

/// @title EIP-4824 DAOs
/// @dev See <https://eips.ethereum.org/EIPS/eip-4824>

interface EIP4824RegistrationFactory is CloneFactory {
    function summonRegistration();
}
```

### Members

Members JSON-LD Schema.

```json

> {
>   "@context": "<http://www.daostar.org/schemas>",
>   "type": "DAO",
>   "name": "<name of the DAO>",
>   "members": [
>     {
>       "type": "EthereumAddress",
>       "id": "<address or other identifier>"
>     },
>     {
>       "type": "EthereumAddress",
>       "id": "<address or other identifier>"
>     }
>   ]
> }

Reviewer 2: ```suggestion
{
  "@context": "<http://www.daostar.org/schemas>",
  "type": "DAO",
  "name": "<name of the DAO>",
  "members": [{
    "type": "EthereumAddress",
    "id": "<address or other identifier>"
  }, {
    "type": "EthereumAddress",
    "id": "<address or other identifier>"
  }]
}
```
My personal preference on how JSON should be formatted. This should probably be standardized.

```

### Proposals

> Proposals JSON-LD Schema. Every EIP-4824 contract should implement a proposalsURI pointing to a JSON object satisfying this schema.

Reviewer 2: ```suggestion
Proposals JSON-LD Schema. Every contract implementing this EIP should implement a proposalsURI pointing to a JSON object satisfying this schema.
```

In particular, any on-chain proposal MUST be associated to an id of the form CAIP10_ADDRESS + “?proposalId=” + PROPOSAL_COUNTER, where CAIP10_ADDRESS is an address following the CAIP-10 standard and PROPOSAL_COUNTER is an arbitrary identifier such as a uint256 counter or a hash that is locally unique per CAIP-10 address. Off-chain proposals MAY use a similar id format where CAIP10_ADDRESS is replaced with an appropriate URI or URL.

```json
{
  "@context": "http://www.daostar.org/schemas",
  "type": "DAO",
  "name": "<name of the DAO>",
  "proposals": [
    {
      "type": "proposal",
      "id": "<proposal ID>",
      "name": "<name or title of proposal>",
      "contentURI": "<URI to content or discussion>",
      "status": "<status of proposal>",
      "calls": [
        {
	  "type": "CallDataEVM",
	  "operation": "<call or delegate call>",
          "from": "<EthereumAddress>",
          "to": "<EthereumAddress>",
	  "value": "<value>",
          "data": "<call data>"
        }
      ]
    }
  ]

> }

Reviewer 5: Note: I *do* support at least internal consistency within a given document.  This file is a mix of tabs and spaces that at least should be addressed, even if the author prefers their other stylistic choices.

> }

Author: fixed; changed everything to 4 spaces. since JSON style is not a blocker, I reverted the [ { line format to the one we've been using throughout our docs. I find it slightly preferable

```

### Activity Log

Activity Log JSON-LD Schema.

```json

> {
>   "@context": "<http://www.daostar.org/schemas>",
>   "type": "DAO",
>   "name": "<name of the DAO>",
>   "activities": [
>     {
>       "id": "<activity ID>",
>       "type": "activity",
>       "proposal": {
>         "id": "<proposal ID>",
>         "type": "proposal"
>       },
>       "member": {
>         "type": "EthereumAddress",
>         "id": "<address or other identifier>"
>       }
>     },
>     {
>       "id": "<activity ID>",
>       "type": "activity",
>       "proposal": {
>         "id": "<proposal ID>",
>         "type": "proposal"
>       },
>       "member": {
>         "type": "EthereumAddress",
>         "id": "<address or other identifier>"
>       }
>     },
>   ]
> }

Reviewer 2: ```suggestion
{
  "@context": "<http://www.daostar.org/schemas>",
  "type": "DAO",
  "name": "<name of the DAO>",
  "activities": [{
    "id": "<activity ID>",
    "type": "activity",
    "proposal": {
      "id": "<proposal ID>",
      "type": "proposal"
    },
    "member": {
      "type": "EthereumAddress",
      "id": "<address or other identifier>"
    }
  }, {
    "id": "<activity ID>",
    "type": "activity",
    "proposal": {
      "id": "<proposal ID>",
      "type": "proposal"
    },
    "member": {
      "type": "EthereumAddress",
      "id": "<address or other identifier>"
    }
  }]
}
```

We chose the word “governance” as an appropriate word that reflects (1) the widespread use of the word in the DAO ecosystem and (2) the common practice of emitting a governance.md file in open-source software projects.

*Alternative names considered: description, readme, constitution*

### Why JSON-LD

We chose to use JSON-LD rather than the more widespread and simpler JSON standard because (1) we want to support use-cases where a DAO wants to include members using some other form of identification than their Ethereum address and (2) we want this standard to be compatible with future multi-chain standards. Either use-case would require us to implement a context and type for addresses, which is already implemented in JSON-LD.

> Further, given the emergence of patterns such as subDAOs and DAOs of DAOs in large organizations such as Synthetix, as well as L2 and multi-chain use-cases, we expect some organizations will point multiple EIP-4824 DAOs to the same URI, which would then serve as a gateway to data from multiple contracts and services. The choice of JSON-LD allows for easier extension and management of that data.

Reviewer 2: ```suggestion
Further, given the emergence of patterns such as subDAOs and DAOs of DAOs in large organizations such as Synthetix, as well as L2 and multi-chain use-cases, we expect some organizations will point multiple DAOs to the same URI, which would then serve as a gateway to data from multiple contracts and services. The choice of JSON-LD allows for easier extension and management of that data.
```

### **Community Consensus**

> The initial draft standard was developed as part of the DAOstar One roundtable series, which included representatives from all major EVM-based DAO frameworks (Aragon, Compound, DAOstack, Gnosis, Moloch, OpenZeppelin, and Tribute), a wide selection of DAO tooling developers, as well as several major DAOs. Thank you to all the participants of the roundtable, the full list of which can be found [here](https://daostar.one/). We would especially like to thank Auryn Macmillan, Fabien of Snapshot, Selim Imoberdorf, Lucia Korpas, and Mehdi Salehi for their contributions.

Reviewer 2: ```suggestion
The initial draft standard was developed as part of the DAOstar One roundtable series, which included representatives from all major EVM-based DAO frameworks (Aragon, Compound, DAOstack, Gnosis, Moloch, OpenZeppelin, and Tribute), a wide selection of DAO tooling developers, as well as several major DAOs. Thank you to all the participants of the roundtable. We would especially like to thank Auryn Macmillan, Fabien of Snapshot, Selim Imoberdorf, Lucia Korpas, and Mehdi Salehi for their contributions.
```
No external links. Feel free to include this in the assets folder, however!

status: Draft
type: Standards Track
category: ERC
created: 2022-02-17
---

## Abstract

An API standard for decentralized autonomous organizations (DAOs), focused on relating on-chain and off-chain representations of membership and proposals.

## Motivation

DAOs, since being invoked in the Ethereum whitepaper, have been vaguely defined. This has led to a wide range of patterns but little standardization or interoperability between the frameworks and tools that have emerged. Standardization and interoperability are necessary to support a variety of use-cases. In particular, a standard daoURI, similar to tokenURI in [EIP-721](./eip-721), will enhance DAO discoverability, legibility, proposal simulation, and interoperability between tools. More consistent data across the ecosystem is also a prerequisite for future DAO standards.

## Specification

The key words “MUST”, “MUST NOT”, “REQUIRED”, “SHALL”, “SHALL NOT”, “SHOULD”, “SHOULD NOT”, “RECOMMENDED”, “MAY”, and “OPTIONAL” in this document are to be interpreted as described in RFC 2119.

Every contract implementing this EIP MUST implement the `EIP4824` interface below:

```solidity

> pragma solidity ^0.4.20;

Reviewer 2: 😮

Author: Is it the pragma?

Author: 😆 I upped it to 0.8.1 to make it consistent with the other solidity code in the standard.
The pragma above was originally copied from ERC 721.

    "type": "DAO",
    "name": "<name of the DAO>",
    "proposals": [
        {
            "type": "proposal",
            "id": "<proposal ID>",
            "name": "<name or title of proposal>",
            "contentURI": "<URI to content or discussion>",
            "status": "<status of proposal>",
            "calls": [
                {
                    "type": "CallDataEVM",
                    "operation": "<call or delegate call>",
                    "from": "<EthereumAddress>",
                    "to": "<EthereumAddress>",
                    "value": "<value>",
                    "data": "<call data>"
                }
            ]
        }
    ]
}
```

### Activity Log

Activity Log JSON-LD Schema.

```json
{
    "@context": "<http://www.daostar.org/schemas>",
    "type": "DAO",
    "name": "<name of the DAO>",
    "activities": [
        {
            "id": "<activity ID>",
            "type": "activity",
            "proposal": {
                "id": "<proposal ID>",
                "type": "proposal"
            },
            "member": {
                "type": "EthereumAddress",
                "id": "<address or other identifier>"
            }

>         }, 

Reviewer 2: ```suggestion
        }, 
      ],
```

>     "activities": [

Reviewer 5: ```suggestion
    ],
    "activities": [
```
